### PR TITLE
some bug fixes with scoping same-named macros

### PR DESCRIPTION
**Pull request by Gvozd**
Also for fix this troubles i som e refactoring main traverse logic
Please look diagram at src/main.odg

if this changes ok, please merge, and publish package
You forgot publish npm-package in previous time

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -178,65 +206,73 @@ export default function build (babel: Object): Object {
...
+  function getMacro (node: Object, scope: Object, state: Object): boolean {
     if (node.type === 'CallExpression') {
-      return getMacro(node.callee, scope);
+      return getMacro(node.callee, scope, state);
     }
     else if (t.isIdentifier(node)) {
-      while(scope) {
-        if (scope._registeredMacros && scope._registeredMacros[node.name]) {
-          return scope._registeredMacros[node.name];
+      if(state._codemixMacros.macrosDefined) {
```
Comment (phpnode) [inline src/index.js]: Let's use a `Symbol('hasMacros')` here

**Comment by phpnode** [comment]
@Gvozd I cleaned up the error messages a little bit, switched to Symbols for private fields and avoided a `traverse()`, please take a look.

**Comment by phpnode** [comment]
@Gvozd do you have an npm username? I tried Gvozd but couldn't find a user. If you create an account I'll give you `npm publish` access.

**Comment by Gvozd** [comment]
I did not know about the thing path.get() )
Your change are OK

Please give me some links for babel-plugin's development API?
I found only babel-handbook, and API's, which I noticed in debugging

Also, i just registered into https://www.npmjs.com/~gvozd

**Comment by phpnode** [comment]
@Gvozd the docs are slim, the only thing to do is to read through [all of this code](https://github.com/babel/babel/tree/master/packages/babel-traverse/src), which luckily is not *so* big.

There are a lot utility functions in there which will make this library faster/better, particularly around scopes. I notice that the `map-filter` test is really slow and we're probably doing a lot of unnecessary work  so there's plenty of room for improvement :)

Thanks for your work on this, feel free to publish those missing tags and this latest work to npm, I just added you as a collaborator.

**Comment by Gvozd** [comment]
Thanks
I already publish package & fix missed tags

My next plan:
1. write more tests, and fix bugs for him, if has.
I have some tricky tests in my losal stash and notes
2. Refactor for readability
Only when, i have all needed tests
3. function inlining for macro arguments
4. Performance
5. ...

**Pull request merged**